src/condor_startd.V6/ResMgr.cpp

```diff
@@ -1711,23 +1681,64 @@ void ResMgr::assign_load_and_idle()
 			total_owner_load);
 	}
 
-	// Distribute the owner load over the slots, assign an owner load of 1.0
-	// to each slot until the remainer is less than 1.0.  then assign the remainder
+	// Distribute the owner load over the slots, assign an owner load equal to Cpus
+	// to each slot until the remainder is less than 1.0.  then assign the remainder
 	// to the next slot, and 0 to all of the remaining slots.
-	// Note that before HTCondor 10.x we would assign *all* of the remainder to the last slot
-	// even if the value was greater than 1.0, but other than that this algorithm is
-	// the same as before.  This algorithm doesn't make a lot of sense for multi-core slots
-	// but it's the way it has always worked so...
-        for (Resource* rip : active) {
-                long long cpus = rip->r_attr->num_cpus();
-                if (total_owner_load < cpus) {
-                        rip->set_owner_load(total_owner_load);
-                        total_owner_load = 0;
-                } else {
-                        rip->set_owner_load(cpus);
-                        total_owner_load -= cpus;
-                }
-        }
+	// The owner load is split up between the slots in slot order, with d-slots
+	// being given the same load as their parent limited by the d-slot core count. 
+	// Before 24.0 the order was Owner, Unclaimed, Matched, Claimed, Preempting
+	// But starting with 24.0 the order is Owner, Unclaimed, <all-other-states>
+	
+	// First distribute load to slots in owner state (usually there aren't any)
+	for (Resource* rip : slots) {
+		if ( ! rip || rip->is_broken_slot() || rip->is_dynamic_slot()) continue;
+		if (rip->state() == State::owner_state)	{
+			total_owner_load = distribute_load(rip, total_owner_load);
+		}
+	}
+
+	// Now distribute load to slots in unclaimed state
+	for (Resource* rip : slots) {
+		if ( ! rip || rip->is_broken_slot() || rip->is_dynamic_slot()) continue;
+		if (rip->state() == State::unclaimed_state)	{
+			total_owner_load = distribute_load(rip, total_owner_load);
+		}
+	}
+
+	// Now distribute d-slot load and load to slots that are not owner or unclaimed
+	// The slots vector puts d-slots after their parent p-slot so
+	// we don't need a separate loop for dslots
+	for (Resource* rip : slots) {
```

Comment:
I feel like we could be a bit more efficient looping the resources rather than looping 3 times... but I am not going to hold this PR back

Reply:
not as bad as you think.   walking a vector of < 1000 elements is pretty fast, the old code allocated a copy of the vector and sorted it, which is almost certainly worse than walking the vector 3 times.

Reply:
Fair point